File: Algorithms/Swift/Easy/19_Remove_Nth_Node_From_End_of_List.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func removeNthFromEnd(head: ListNode?, _ n: Int) -> ListNode? {
        if head == nil || n <= 0 {
            return head
        }
    
        let dummy = ListNode(0)
        dummy.next = head
        
        var node: ListNode? = dummy
        for i in 0 ..< n {
            if node == nil {
                return head
            }
            node = node!.next
        }
        
        var tail = node
        node = dummy
        while tail!.next != nil {
            node = node!.next
            tail = tail!.next
        }
        
        let nextNode = node!.next
        node!.next = nextNode!.next
        
        return dummy.next
    }
}

File: Algorithms/JavaScript/Easy/19_Remove_Nth_Node_From_End_of_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    if (head === null || n <= 0) {
        return head
    }

    var dummy = new ListNode(0)
    dummy.next = head

    var node = dummy
    for (var i = 0; i < n; i++) {
        if (node === null) {
            return head
        }
        node = node.next
    }

    var tail = node
    node = dummy
    while (tail.next !== null) {
        node = node.next
        tail = tail.next
    }

    var nextNode = node.next
    node.next = nextNode.next

    return dummy.next

};
